### PR TITLE
[22.05] New Datatype: QIIME 2

### PR DESCRIPTION
**Pull request by ebolyen**
Initial support for QIIME 2. We've used this successfully in our most recent workshop. 

These changes are used by [q2galaxy](https://github.com/qiime2/q2galaxy), the integration of which is tested by [q2-mystery-stew](https://github.com/qiime2/q2-mystery-stew) (though more focused tests should be written here as well), and will support our [toolshed repo](https://github.com/qiime2/galaxy-tools).

As I'm coming back to this, I am pretty sure there's a bit of lint to clean up. In particular I think the column type conversion between our "QIIME 2 Metadata" file and Galaxy Tabular formats is not at 100% yet. I'd like to see something more robust here.

## How to test the changes?
(Select all options that apply)
- [x] TODO: I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Install the latest [staging environment for QIIME 2](https://dev.qiime2.org/latest/quickstart/#install-qiime-2-within-a-conda-environment)
  2. Run `q2galaxy template tests some_directory/`
  3. Use planemo to test the resulting tools in the usual way (which should take about an hour or so to run).
 
## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by ebolyen** [comment]
I'm going to take a fresh pass at some of the specifics here, but would appreciate any feedback on the overall structure.

**Comment by ebolyen** [comment]
Test failures are from a function rename, so that makes sense.

File: test/unit/data/datatypes/test_qiime2.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        dataset = MockDataset(1)
+        dataset.file_name = input_files[0]
+
+        qza.set_meta(dataset)
+
+        assert dataset.metadata.uuid == 'ba8c55e1-a2bc-47ea-beb2-b37b0b3b4032'
+        assert dataset.metadata.version == '2022.2.1'
+        assert dataset.metadata.format == 'SingleIntDirectoryFormat'
+        assert dataset.metadata.semantic_type == 'SingleInt1'
+        assert dataset.metadata.semantic_type_simple == 'SingleInt1'
```
Comment (ebolyen) [inline test/unit/data/datatypes/test_qiime2.py]: mypy doesn't like these tests on the mock, what should I do in this case?

**Comment by ebolyen** [comment]
Results from running the following within a [QIIME 2 environment](https://docs.qiime2.org/2022.2/install/native/) (after pip installing planemo):

```
q2galaxy template tests ./rendered/tests/
```

```
planemo test --install_galaxy \
  --galaxy_branch initial_qiime2_support \
  --galaxy_source https://github.com/ebolyen/galaxy.git \
  --no_conda_auto_install \
  --no_conda_auto_init \
  --test_output ./rendered/tests/tool_test_output.html \
  --test_output_json ./rendered/tests/tool_test_output.json \
  ./rendered/tests/suite_qiime2_mystery-stew/
```

[image planemo_results.zip](https://github.com/galaxyproject/galaxy/files/9015041/planemo_results.zip)

**Comment by nsoranzo** [comment]
@ebolyen `make format` should fix (most of) the Python linting errors.

**Comment by ebolyen** [comment]
Thanks @nsoranzo! I've got a local commit with those changes, just don't want to waste your CI until I have the next part figured out:

---

Related to the integration tests, I think I must be doing something really quite wrong.

I'm seeing this error (and a similar one for qzv, the tabular one I think is just because I used the wrong extension)
```
____________________ test_upload_datatype_auto[qiime2.qza] _____________________
instance = <integration.test_datatype_upload.UploadTestDatatypeDataTestCase object at 0x7f5e2530cad0>
test_data = TestData(path='/home/runner/work/galaxy/galaxy/galaxy root/test/integration/../../lib/galaxy/datatypes/test/qiime2.qza', datatype=<galaxy.datatypes.qiime2.QIIME2Artifact object at 0x7f5e02e64b90>, uploadable=True)
temp_file = <tempfile._TemporaryFileWrapper object at 0x7f5e251f5a10>
celery_session_worker = <Worker: gen7888@fv-az108-206 (running)>
celery_session_app = <Celery celery.tests at 0x7f5e001f48d0>
    @pytest.mark.parametrize("test_data", TEST_CASES.values(), ids=list(TEST_CASES.keys()))
    def test_upload_datatype_auto(instance, test_data, temp_file, celery_session_worker, celery_session_app):
>       upload_datatype_helper(instance, test_data, temp_file)
test/integration/test_datatype_upload.py:60: 
```
<details><summary>Middle of the pytest stack trace</summary>

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
instance = <integration.test_datatype_upload.UploadTestDatatypeDataTestCase object at 0x7f5e2530cad0>
test_data = TestData(path='/home/runner/work/galaxy/galaxy/galaxy root/test/integration/../../lib/galaxy/datatypes/test/qiime2.qza', datatype=<galaxy.datatypes.qiime2.QIIME2Artifact object at 0x7f5e02e64b90>, uploadable=True)
temp_file = <tempfile._TemporaryFileWrapper object at 0x7f5e251f5a10>
delete_cache_dir = False
    def upload_datatype_helper(instance, test_data, temp_file, delete_cache_dir=False):
        is_compressed = False
        for is_method in (is_bz2, is_gzip, is_zip):
            is_compressed = is_method(test_data.path)
            if is_compressed:
                break
        with open(test_data.path, "rb") as content:
            if hasattr(test_data.datatype, "sniff") or "false" in test_data.path:
                file_type = "auto"
            else:
                file_type = test_data.datatype.file_ext
            dataset = instance.dataset_populator.new_dataset(
                instance.history_id,
                content=content,
                wait=False,
                file_type=file_type,
                auto_decompress=True,
            )
        dataset = instance.dataset_populator.get_history_dataset_details(
            instance.history_id, dataset=dataset, assert_ok=False
        )
        expected_file_ext = test_data.datatype.file_ext
        # State might be error if the datatype can't be uploaded
        if dataset["state"] == "error" and not test_data.uploadable:
            # Some things can't be uploaded, if attempting to upload these datasets we mention why
            assert "invalid" in dataset["misc_info"] or "unsupported" in dataset["misc_info"]
            return
        elif dataset["state"] == "error" and "empty" in dataset["misc_info"]:
            return
        else:
            # state should be OK
            assert dataset["state"] == "ok"
        # Check that correct datatype has been detected
        file_ext = dataset["file_ext"]
        if "false" in test_data.path:
            # datasets with false in their name are not of a specific datatype
            assert file_ext != PARENT_SNIFFER_MAP.get(expected_file_ext, expected_file_ext)
        else:
```
</details>

```
>           assert file_ext == PARENT_SNIFFER_MAP.get(expected_file_ext, expected_file_ext)
E           AssertionError: assert 'txt' == 'qza'
E             - qza
E             + txt
```

So it appears that the upload test with format "auto" is deciding this is a txt file. Which is strange, because I was pretty sure that I have the order of the sniffers for qza preceding the more general zip archive. Also strange is that I don't see this behavior normally, just in the test.

---

I also can't figure out how to reproduce this effectively as I get timeouts with this command:
```
 ./run_tests.sh -integration test/integration/test_datatype_upload.py::test_upload_datatype_auto[qiime2.qza]
```
Pytest collects it just fine, but I imagine some fixture wasn't set up so I get this error locally:

```
galaxy.tool_util.verify.wait.TimeoutAssertionError: Timed out after 60.25 seconds waiting on history state. Current response containing state [{'model_class': 'History', 'id': 'adb5f5c93f827949', 'name': 'API Test History', 'deleted': False, 'purged': False, 'url': '/api/histories/adb5f5c93f827949', 'published': False, 'annotation': None, 'tags': [], 'contents_url': '/api/histories/adb5f5c93f827949/contents', 'size': 0, 'user_id': 'adb5f5c93f827949', 'create_time': '2022-06-30T17:17:46.571146', 'update_time': '2022-06-30T17:17:47.000447', 'importable': False, 'slug': None, 'username_and_slug': None, 'genome_build': None, 'state': 'queued', 'state_ids': {'new': [], 'upload': [], 'queued': ['adb5f5c93f827949'], 'running': [], 'ok': [], 'empty': [], 'error': [], 'paused': [], 'setting_metadata': [], 'failed_metadata': [], 'deferred': [], 'discarded': []}, 'state_details': {'new': 0, 'upload': 0, 'queued': 1, 'running': 0, 'ok': 0, 'empty': 0, 'error': 0, 'paused': 0, 'setting_metadata': 0, 'failed_metadata': 0, 'deferred': 0, 'discarded': 0}, 'hid_counter': 2, 'empty': False}].
```

That's just user error on my part of course. The galaxy server does seem to be active and even has a job:
```
celery.worker.request DEBUG 2022-06-30 10:17:47,350 [pN:main,p:17069,tN:Thread-2] Task accepted: galaxy.celery.tasks.setup_fetch_data[1801bf58-bc69-4eba-8326-68bb5397baa2] pid:17069
galaxy.jobs DEBUG 2022-06-30 10:17:47,423 [pN:main,p:17069,tN:Thread-2] (1) Working directory for job is: /tmp/tmp3t9ozr6l/tmptfa3_2v2/tmpn3tovy27/database/job_working_directory_60u23n1j/000/1
INFO:     127.0.0.1:45906 - "GET /api/histories/adb5f5c93f827949 HTTP/1.1" 200 OK
galaxy.jobs DEBUG 2022-06-30 10:17:47,577 [pN:main,p:17069,tN:Thread-2] Job wrapper for Job [1] prepared (40.038 ms)
galaxy.celery INFO 2022-06-30 10:17:47,577 [pN:main,p:17069,tN:Thread-2] Successfully executed Celery task setup_fetch_data setup_fetch_data (226.659 ms)
INFO:     127.0.0.1:45908 - "GET /api/histories/adb5f5c93f827949 HTTP/1.1" 200 OK
```

---

Thanks to writing all of that out, I have a few ideas I'm going to check on, such as some exception getting raised in this context causing the auto-detection to skip me. 

Or perhaps I've violated some rule about how these test data are supposed to be placed/registered.

But if anyone else has ideas, I would appreciate any help.

**Comment by ebolyen** [comment]
Update: my schemas were old and bad `git clean -fxd` did the tick, I can now see this locally and running as a server.

Seems that something really has changed in the last year. But now I have something I can debug.

**Comment by ebolyen** [comment]
Ok, starting to figure this out:

The `sniff()` methods aren't being called on the raw zip archive because we are ending up in this branch:
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/sniff.py#L683-L690

It looks like the compressed file handling system was overhauled in https://github.com/galaxyproject/galaxy/pull/13654 but I don't see immediately what I need to do to get our data-types back in line with it. We are inheriting from `CompressedZipArchive`, but it may be the case that that is also broken.

**Comment by ebolyen** [comment]
I was able to fix it by adding `is_compressed = "maybe"` to `CompressedZipArchive` in [b647b3a](https://github.com/galaxyproject/galaxy/pull/14216/commits/b647b3ab585b358a2cec33f367f1b58df6462665).
I think that is technically true, and it seems to restore the original behavior of that parent class w.r.t. the upload tool.

We'll see if the tests agree (I haven't been able to get around the timeout locally yet).

**Comment by ebolyen** [comment]
Ok, I don't think that selenium failure is related to this PR, but I was really hoping this would be the one to go all green.

**Review by mvdbeek — APPROVED** [review]
Looks great, thanks a lot @ebolyen!

**Review by bgruening — APPROVED** [review]
Möp! This is pretty cool! Looks great to me and the visualizations, awesome stuff!

@ebolyen I would encourage you to change the target branch to 22.05. For datatypes we are allowed to add them even after branching.

**Comment by ebolyen** [comment]
will rebase this in just a moment

**Comment by ebolyen** [comment]
I've rebased on `release_22.05`, everything looks right here, so we'll see if CI agrees.

**Comment by ebolyen** [comment]
Question about the process from this point, is the `release_22.05` branch "forward-ported" to `dev` periodically, or do we set up parallel commit histories from this point on?

**Comment by nsoranzo** [comment]
> Question about the process from this point, is the `release_22.05` branch "forward-ported" to `dev` periodically (with some merge commit), or do we set up parallel commit histories from this point on?

Release branches are manually but regurlarly merged forward, e.g. release_21.09 -> release_22.01 -> release_22.05 -> dev . Also the master branch is kept in sync with the latest stable release branch (currently release_22.01).

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Very cool @ebolyen!

**Comment by ebolyen** [comment]
Thanks for the reviews @mvdbeek and @bgruening! And thanks for the Galaxy advice @nsoranzo!

Very excited to get our toolshed in working order next!

**Comment by gregvonkuster** [comment]
This is awesome - thanks so much for all of this work!  I'm looking forward to the QIIME 2 tools!  :)

**Pull request merged**